Analyze tests directory with psalm

franmomu opened the pull request:
<!-- Fill in the relevant information below to help triage your pull request. -->

|      Q       |   A
|------------- | -----------
| Type         | improvement
| BC Break     | no
| Fixed issues | <!-- use #NUM format to reference an issue -->

#### Summary

<!-- Provide a summary your change. -->

Adding static analysis to `tests` directory allow us to spot some issues.

In this case with psalm since with phpstan there are quite a lot issues, I'll be creating PRs to try to fix them little by little (there were more than 300).

franmomu commented on tests/Doctrine/ODM/MongoDB/Tests/Functional/TestTargetDocument.php:
```diff
@@ -30,8 +30,15 @@ class TargetDocumentTestDocument
+/**
+ * @ODM\Document
+ */
+class TargetDocumentTestReference extends AbstractTargetDocumentTestReference
```
It was instantiating an abstract class in https://github.com/doctrine/mongodb-odm/pull/2311/files#diff-6f1c94e0061bd56051708e36273680c748d5dba9ac84f2cc319533a46ac78063R15

alcaeus replied:
The link doesn't go anywhere for me, could you give me a link? This might be a bug since we shouldn't try to instantiate abstract classes.

franmomu replied:
Oops sorry, as I've split them into commits I guess it got lost, here: https://github.com/doctrine/mongodb-odm/blob/f225a0cfc5ba878aba3e32b7f5fd70facc2d4eee/tests/Doctrine/ODM/MongoDB/Tests/Functional/TestTargetDocument.php#L15

alcaeus replied:
Ok, I'll take a look and see what that test was doing. Thanks!

alcaeus replied:
Yes, that change makes absolute sense, thank you!

franmomu commented on tests/Doctrine/ODM/MongoDB/Tests/Performance/UnitOfWorkPerformanceTest.php:
```diff
@@ -22,8 +22,9 @@ class UnitOfWorkPerformanceTest extends BaseTest
      */
     public function testComputeChanges()
     {
+        $n     = 10000;
```
`$n` is used in https://github.com/doctrine/mongodb-odm/pull/2311/files#diff-f3c0f4bbc68f314a5f5a2f8c28f67b508f8bb9c452f14ab29203a2a01cc28f15R48

franmomu commented on tests/Documents/Bars/Location.php:
```diff
@@ -17,11 +17,6 @@ public function __construct($name = null)
 
-    public function getId()
-    {
-        return $this->id;
```
`id` property is not in the class.

franmomu committed:
Use int for constructing UTCDateTime

franmomu committed:
Add assertions

franmomu committed:
Add missing attributes

franmomu committed:
Fix phpunit class name

franmomu committed:
Use real class instead of generating it

franmomu committed:
Do not instantiate abstract class

franmomu committed:
Fix param names

franmomu committed:
Fix property and return types

franmomu committed:
Use existing classes

franmomu committed:
Remove methods accessing to unexisting properties

franmomu committed:
Fix unexisting variable

franmomu committed:
Remove dead code

franmomu committed:
Add static analysis for tests with psalm

franmomu commented:
I got caught up making changes 😬 , I've tried to split it in commits and just change _simple_ things that should be easy to review, otherwise let me know and I can split it in PRs.

After this there are _just_ 130 remaining issues in phpstan in tests.

alcaeus commented on tests/Doctrine/ODM/MongoDB/Tests/Functional/EcommerceTest.php:
```diff
@@ -78,6 +78,7 @@ public function testRemoveOption()
         $this->dm->flush();
         $this->dm->detach($product);
         unset($product);
+        /** @psalm-suppress UndefinedVariable */
```
I'm not sure why the assertion even has to happen. is this test trying to test PHP behaviour? IMO we can the `assertFalse` call instead.

alcaeus commented on tests/Doctrine/ODM/MongoDB/Tests/Functional/EmbeddedTest.php:
```diff
@@ -112,6 +113,7 @@ public function testOneEmbedded()
             ->field('id')->equals($user->getId())
             ->getQuery()
             ->getSingleResult();
+        assert($user instanceof User);
```
This shows me that we're still missing some psalm template magic. I'll see if I can fix that somehow.

franmomu replied:
Yeah, exactly, it would be great to be able to remove those and eventually I hope we will. `Builder` and `Query` should be generic I guess, but I didn't try yet. (**Edit**: When generics would be added, psalm would complain about this being always true.)

I've also added template to [`Cursor` and `CursorInterface`](https://github.com/vimeo/psalm/blob/master/stubs/mongodb.phpstub) just in case we want to go deeper.

alcaeus commented on tests/Doctrine/ODM/MongoDB/Tests/UnitOfWorkTest.php:
```diff
@@ -518,7 +518,6 @@ public function testCommitsInProgressIsUpdatedOnException()
         $this->expectExceptionMessage('This should not happen');
 
         $this->dm->flush();
-        $this->assertAttributeSame(0, 'commitsInProgress', $this->dm->getUnitOfWork());
```
Can we replace this assertion in a different way, maybe by accessing the attribute in a bound closure?

franmomu replied:
Hmmm the thing about this line is that it never gets executed because the `flush()` method throws an exception, but reading the test name: `testCommitsInProgressIsUpdatedOnException` maybe that exception should be catch and check the `commitsInProgress` property?

alcaeus replied:
On second thought, yes. The method should catch the exception, check the `commitsInProgress` property and then return. If no exception was caught, the test should fail using `$this->fail()`.

franmomu replied:
perfect, done.

alcaeus reviewed (approved):
Great work. You mentioned there being a number of errors left. How would one go about fixing those?

franmomu committed:
Remove unnecessary assertion

franmomu committed:
Temporary suppress psalm issues with toPHP method

There should be added to the mongodb stub of psalm, we are using
findUnusedPsalmSuppress configuration option, so it will warn us
when this is fixed.

franmomu committed:
Check commitsInProgress in UnitOfWork

franmomu committed:
Fix phpdoc

franmomu committed:
Add missing class properties

franmomu commented:
> You mentioned there being a number of errors left. How would one go about fixing those?

I've created https://github.com/doctrine/mongodb-odm/pull/2317

alcaeus commented on tests/Doctrine/ODM/MongoDB/Tests/Mapping/ClassMetadataTest.php:
```diff
@@ -64,6 +64,7 @@ public function testClassMetadataInstanceSerialization()
         $cm->setVersioned(true);
         $cm->setVersionField('version');
         $validatorJson = '{ "$and": [ { "email": { "$regex": { "$regularExpression" : { "pattern": "@mongodb\\\\.com$", "options": "" } } } } ] }';
+        /** @psalm-suppress TooFewArguments */
```
What's the cause for this error? Is the documentation here incorrect or is it the code that's wrong?

franmomu replied:
The code is fine, I think the problem is in Psalm that doesn't have a stub specifying that the second parameter is optional. I've created https://github.com/vimeo/psalm/pull/5864

alcaeus reviewed (commented):
One question about the new `@psalm-suppress`, but the rest looks good!

alcaeus commented:
Thanks @franmomu!
